src/Common/IFileCachePriority.h

```diff
@@ -0,0 +1,95 @@
+#pragma once
+
+#include <memory>
+#include <mutex>
+#include <Core/Types.h>
+#include <Common/Exception.h>
+#include <Common/FileCacheType.h>
+
+namespace DB
+{
+
+class IFileCachePriority;
+using FileCachePriorityPtr = std::shared_ptr<IFileCachePriority>;
+
+/// IFileCachePriority is used to maintain the priority of cached data.
+class IFileCachePriority
+{
+public:
+    class IIterator;
+    using Key = FileCacheKey;
+    using ReadIterator = std::shared_ptr<const IIterator>;
+    using WriteIterator = std::shared_ptr<IIterator>;
```

Comment:
may be unique_ptr?

Reply:
ReadIterator can be unique_ptr. But for WriteIterator, I don't know if it should be set to unique_ptr, because it may be copied with the FileSegment, do we need to invalidate the iterator before copying?

Reply:
>  because it may be copied with the FileSegment, do we need to invalidate the iterator before copying?

but file segment cannot be copied

Reply:
It looks like this, I tried to modify it, but sometimes we need to use a temporary pointer to keep some code readable, such as the following cases, so I think it may be better to keep it as it is, what do you think?
```
auto priority_iter = record->second;
priority_iter->use(cache_lock);
result_state = priority_iter->hits() >= enable_cache_hits_threshold ? FileSegment::State::EMPTY : FileSegment::State::SKIP_CACHE;
```

Reply:
but looks like here `auto & ` is possible

Reply:
I tried to fix it, but I always get an error when compiling involves unordered_map. Maybe there is something wrong with my understanding of uniqueu_ptr :(

Reply:
may be not use a pointer at all - just an object? (and in FileSegmentCell put std::optional to iterator).

Reply:
Sounds feasible, do you mean all ReadIterator and WriteIterator as objects? In this case, there will be more revisions, but I may not have much time to make revisions recently (there is a lot of work to deal with in reality), and I will reconsider the relevant implementation when I have time :).

Reply:
> do you mean all ReadIterator and WriteIterator as objects

yes

> but I may not have much time to make revisions recently (there is a lot of work to deal with in reality), and I will reconsider the relevant implementation when I have time

ok, let's leave it as it is for now